iree/compiler/Conversion/LinalgToLLVM/ConvImg2ColMatmulConversion.cpp

```diff
@@ -0,0 +1,180 @@
+ public:
+  using OpRewritePattern<linalg::ConvOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(linalg::ConvOp op,
+                                PatternRewriter &rewriter) const override {
+    auto filterShapeType = op.filter().getType().dyn_cast_or_null<ShapedType>();
+    auto inputShapeType = op.input().getType().dyn_cast_or_null<ShapedType>();
+    auto outputrShapeType =
+        op.output().getType().dyn_cast_or_null<ShapedType>();
+    if (!filterShapeType || !inputShapeType) return failure();
+
+    auto loc = op.getLoc();
+
+    auto inputFeatures =
+        filterShapeType.getShape()[filterShapeType.getRank() - 1];
+    auto outputFeatures = filterShapeType.getShape().back();
+
+    // Col buffer shape (n, d1, d1, d2, ...dn, k1, k2, k3, ...kn, ci)
+    SmallVector<int64_t, 4> colBufferShape;
+    int64_t spatialSize = 1, filterSpatialSize = 1;
+    colBufferShape.push_back(outputrShapeType.getShape()[0]);
+    for (int i = 0; i < op.getNumSpatialDimensions(); ++i) {
+      auto dimSize =
+          outputrShapeType.getShape()[i + op.getNumBatchDimensions()];
+      colBufferShape.push_back(dimSize);
+      spatialSize *= dimSize;
+    }
+    for (int i = 0; i < op.getNumSpatialDimensions(); ++i) {
+      auto dimSize = filterShapeType.getShape()[i];
+      colBufferShape.push_back(dimSize);
+      filterSpatialSize *= dimSize;
+    }
+    colBufferShape.push_back(
+        filterShapeType.getShape()[op.getNumSpatialDimensions()]);
+
+    auto ColBufferMemrefType =
+        MemRefType::get(colBufferShape, filterShapeType.getElementType());
+
+    Value result = rewriter.create<AllocaOp>(loc, ColBufferMemrefType);
+
+    llvm::SmallVector<Value, 4> linalgOpArgs = {op.input(), result};
+
+    // (n, d1, d2, d3, ..., dn, k1, k2, k3, ...kn, ci) ->
+    // (n, d_1 * stride_1 + k_1, d_2 * stride_2 + k_2, ...d_n * stride_n + k_n,
+    // ci)
+    SmallVector<AffineExpr, 4> inputExprs;
+    inputExprs.push_back(rewriter.getAffineDimExpr(0));
+    int spatialDimsOffset = 1;
+    auto kernelDimsOffset = spatialDimsOffset + op.getNumSpatialDimensions();
+    for (int i = 0; i < op.getNumSpatialDimensions(); ++i) {
+      inputExprs.push_back(rewriter.getAffineDimExpr(i + spatialDimsOffset) *
+                               op.getStride(i) +
+                           rewriter.getAffineDimExpr(i + kernelDimsOffset));
+    }
+    inputExprs.push_back(rewriter.getAffineDimExpr(
+        kernelDimsOffset + op.getNumSpatialDimensions()));
+
+    auto nloops = colBufferShape.size();
+
+    SmallVector<StringRef, 3> loopAttributeTypes(nloops, "parallel");
+
+    SmallVector<AffineMap, 4> indexingMaps;
+    indexingMaps.emplace_back(
+        AffineMap::get(nloops, 0, inputExprs, rewriter.getContext()));
+    indexingMaps.emplace_back(AffineMap::getMultiDimIdentityMap(
+        ColBufferMemrefType.getRank(), rewriter.getContext()));
+
+    rewriter.create<linalg::GenericOp>(
```

Comment:
I think you can just use a copy op here with the input permutation and output permutation you already computed.

Reply:
Copy expects only permutations of same rank it wont accept the indexing maps we have here.

Reply:
Oh, didnt realize that. Thanks for checking!